[Chan Wen Yee] iP

chwenyee opened the pull request:
Add some paragraph to the README.md to explain ...
Also add a heading ...

chwenyee committed:
Level-0

chwenyee committed:
Week 3: Level 1. Echo.
Duke echos commands entered by the user, and exits when the user types the command bye.

chwenyee committed:
Level 2. Add, List.
Luke can now store whatever text entered by the user and display them back to the user when requested.

chwenyee committed:
Level 3. Mark as Done.
Duke's new function: the ability to mark tasks as done and the ability to change the status back to not done.

Yshinprograms commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+                    tasks[taskIndex].markAsDone();
+                    System.out.println(DIVIDER);
+                    System.out.println("Nice! I've marked this task as done:");
+                    System.out.println("[" + tasks[taskIndex].getStatusIcon()+ "] " + tasks[taskIndex].description);
```
Maybe long println statements like line 43 and 52 could be shortened or use line wrapping?

Yshinprograms commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+        Scanner in = new Scanner(System.in);
+        input = in.nextLine();
+
+        while (!input.equals("bye")) {
```
Good job following Egyptian style brackets!

Yshinprograms commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+        bye();
+
+    }
+}
```
Indentations throughout the code looks good!

Yshinprograms commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+    public static void main(String[] args) {
+        greeting();
+        //String[] tasks = new String[100];
+        Task[] tasks = new Task[100];
```
Array specifier is attached to the type and not the variable, great!

Yshinprograms reviewed (commented):
LGTM in general, except a few lines seem to be slightly long. Consider splitting them into a few lines or wrap the lines.

Xavierleejrui commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+        int taskCount = 0;
+
+        String input;
+        Scanner in = new Scanner(System.in);
```
Good usage of scanner! However do keep in mind to ensure that when a scanner is declared, it must be closed too!

Xavierleejrui commented on src/main/java/Luke.java:
```diff
@@ -0,0 +1,68 @@
+                    System.out.println("[" + tasks[taskIndex].getStatusIcon()+ "] "+ tasks[taskIndex].description);
+                    System.out.println(DIVIDER);
+                }
+            } else {
```
Good logical flow of the code, it would be good to consider edge cases to make sure the program doesn't crash with invalid commands and account for it.

Xavierleejrui reviewed (commented):
Good implementation so far keep it up

chwenyee committed:
Level 4. ToDos, Events, Deadlines
Duke can now support for tracking three types of tasks such as ToDos, Deadlines and Events given the correct input format.

chwenyee committed:
Level-5. Handle Errors
Add some exception classes that are specific to Luke.

chwenyee committed:
Level 5. Handle Errors.
Luke can now deal with errors such as incorrect inputs entered by the user.

chwenyee committed:
Merge branch 'branch-Level-5'

chwenyee committed:
Level 5 Add packages

chwenyee committed:
Level 5 Set a getter method for taskcount to fix the problem of reaching protected variable.

chwenyee committed:
Merge branch 'branch-A-Packages'

chwenyee committed:
Level 6 Implementation of ArrayList on tasks

chwenyee committed:
Level 6. Delete
Add support for deleting tasks from the list.

chwenyee committed:
Level 7. Save
Luke save the tasks in the hard disk automatically whenever the task list changes. Luke can now load the data from the hard disk when it starts up.

chwenyee committed:
Merge branch 'branch-Level-6'

chwenyee committed:
Merge branch 'branch-Level-7'
Solved conflicts at src/main/java/luke/Luke.java

chwenyee committed:
Improved code. Remove the code that print saved tasks in luke.txt due to misunderstanding.

chwenyee committed:
Resolved error of showing DIVIDER "---------" while using JAR to start Luke up.

chwenyee committed:
Changed the keyword to delete task from "remove" to "delete".

xenthm commented on src/main/java/luke/LukeException.java:
Good job creating sub-exceptions to your custom `LukeException` exception!

xenthm commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,238 @@
+            return String.join(" ", Arrays.copyOfRange(words, filterIndex, indexBetweenFromTo));
+        } else {
+            return String.join(" ", Arrays.copyOfRange(words, filterIndex, words.length));
+        }
```
I like how you use `filterType` as a indicator for what to do in your method, but consider using a constant to represent it to avoid [magic numbers](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers).

xenthm commented on src/main/java/luke/Luke.java:
```diff
@@ -0,0 +1,238 @@
+
+    public static final String DIVIDER = "------------------------------------------------------------------";
+
+    private static void greeting() {
+        System.out.println(DIVIDER);
+        System.out.println("Hi! I'm Luke a.k.a your customized chatbot.");
+        System.out.println("What can I do for you?");
+        System.out.println(DIVIDER);
+    }
+
+    private static void bye() {
+        System.out.println(DIVIDER);
+        System.out.println("It's my pleasure to help you. Hope to see you again soon! :')");
+        System.out.println(DIVIDER);
+    }
```
Great job with naming your variables and methods, but some of them could be improved! The `greeting` and `bye` could be in [verb form](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-nouns-for-things-and-verbs-for-actions).

chwenyee committed:
Make the code more OOP.
Extract the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list
Command classes (i.e., AddCommand, DeleteCommand, ExitCommand etc.) that inherit from an abstract Command class

chwenyee committed:
Updated the more organized code.

chwenyee committed:
Merge branch 'branch-A-MoreOOP'

chwenyee committed:
Some changes on LukeException messages.

chwenyee committed:
Added a few comments in Parser and Storage.

chwenyee committed:
Only added one comment in Storage

chwenyee committed:
Deleted an Exception class that is never used.

chwenyee committed:
Merge pull request #1 from chwenyee/branch-A-Assertions

Merging branch-A-Assertions

chwenyee committed:
Merge branch 'master' into branch-A-CodeQuality

chwenyee committed:
Merge branch 'master' into branch-A-Streams

chwenyee committed:
Merge pull request #3 from chwenyee/branch-A-CodeQuality

Merge branch-A-CodeQuality

chwenyee committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/luke/Storage.java

chwenyee committed:
Merge pull request #2 from chwenyee/branch-A-Streams

Merge branch-A-Streams

chwenyee committed:
Level 9. Find
Luke is now able to find a task by searching for a keyword in the task description provided by users.

chwenyee committed:
Merge branch 'branch-Level-9'

chwenyee committed:
Improved the code.

chwenyee committed:
Improve code quality of Parser

chwenyee committed:
Luke is now able to store deadline dates as a java.time.LocalDateTime in task objects.
It accept dates in a format yyyy-MM-dd HHmm and print in a format MMM dd yyyy, h:mm a.

chwenyee committed:
Level-8 Dates and Times
Luke now understands dates and times. It can also perform query on specified date and print the related tasks for users.

chwenyee committed:
Merge branch 'branch-Level-8'

chwenyee committed:
Updated JavaDoc in a few files.

chwenyee committed:
Added header comments to all non-private classes/methods, and non-trivial private methods.

chwenyee committed:
Merge branch 'branch-A-JavaDoc'

chwenyee committed:
Updated UserGuide in README.md.
Improved code in Parser.java to handle error of loading data with invalid format in text file.

chwenyee committed:
Update README.md in doc

chwenyee committed:
Delete README.md in main

chwenyee committed:
Updated the correct README

chwenyee committed:
Merge branch 'branch-A-JavaDoc'

chwenyee committed:
Revert the version of wrong README that I have edited.

chwenyee committed:
Merge branch 'master' of https://github.com/chwenyee/ip

# Conflicts:
#	README.md
#	docs/README.md

chwenyee committed:
Updated README.md

chwenyee committed:
Updated an image in README.md.

chwenyee committed:
Make LukeException to be in a new package called exception. Updated the displayed message upon receiving invalid command.

chwenyee committed:
Fixed the problem of event, deadline and todo being able to accept blank description. Slight change on README.md.
